### PR TITLE
Added Euler angles rotation orders.

### PR DESCRIPTION
**Pull request by thezwap**
Only Tait-Bryan angles were supported previously.

### CODE REVIEW

**Comment by thezwap** [comment]
Rotations such as ZXZ are very commonly used for calculating part rotations in multi-axis CNC machine controls. I used the maths from the same matlab script that the Tait-Bryan angles were taken from, in the original code (link in comment in Quaternion.js.)

**Comment by mrdoob** [comment]
@WestLangley what do you think?

**Comment by WestLangley** [comment]
@thezwap 

1. Did you test this code?

2. Are the "proper" Euler angles you added of the _intrinsic_ type?

3. There has be no demand for these Euler types in three.js that I have been aware of. Would an `examples/js/math/EulerUtils.js` be acceptable, instead?

4. What would be the minimal set of methods that you would need to add to three.js to satisfy your use case?

Were this to be integrated into core, for completeness, we would also have to add additional conversion methods for `Euler` to-and-from `Quaternion` and `Euler` to-and-from `Matrix4`.

**Comment by thezwap** [comment]
Hi @WestLangley 

1. Yes, I have tested the ZXZ rotation and it works exactly as I’d expect it to. I haven’t tested all the other orders that I added, but I can definitely do that. 

2. Yes, they are intrinsic rotations in their current form. Sometimes we do use extrinsic rotations to suit certain CNC machine controls that expect the tilted workplane to be calculated by a series of rotations around a fixed coordinate system. This is much less common. 

3. I’m kind of surprised that no one has asked for this previously. It’s a very common requirement to calculate the rotations in this way in the world of machine control, and I’ve started to see a few machine control apps built with threejs. See ncviewer.com as an example. That doesn’t do multi axis plane rotations yet as far as I know, but I’d be surprised if they weren’t on the roadmap. 
In my opinion, I feel that for completeness of the Euler library, they are just as valid as the Tait-Bryan rotations. Additionally, it was very confusing to me that I got no rotation at all when I entered “ZXZ” as the order, as the documentation didn’t specify that only Tait-Bryan angles were supported, so it was only when I went digging in the source, that I found out why. I think having them there would save others’ confusion. 

4.  Personally, for my current use case, I’d need all of the methods I’ve added. Each machine control we work with uses a particular order in that set. It varies a lot. 

I could write the conversions to and from quat and matrix4 if necessary. 

Thanks for considering this.

**Comment by WestLangley** [comment]
> I haven’t tested all the other orders that I added

Please always test your PR before your file it.

>as the documentation didn’t specify that only Tait-Bryan angles were supported

The supported rotation orders are specified in http://imac.local/three.js/docs/#api/math/Euler. Unfortunately, it does not appear that `Object3D.rotation` links to that page.

>Personally, for my current use case, I’d need all of the methods I’ve added. Each machine control we work with uses a particular order in that set.

So, what are you doing as a workaround?

**Comment by thezwap** [comment]
> Please always test your PR before your file it.
Fair enough. I'll go through all of them and check they are correct. I've actually just noticed a flaw in my previous method of testing them, so I'll double-check.

> The supported rotation orders are specified in http://imac.local/three.js/docs/#api/math/Euler. 
Ok, I see that now. I think I did actually read that at one point, but misread that as being examples, rather than an exclusive list.

> So, what are you doing as a workaround?
My workaround is the code I've written in a custom build of three.js

**Comment by thezwap** [comment]
When I replied previously and said the maths that I used provided intrinsic rotations, I based that on the fact that reversing the order of rotations, converts intrinsic rotations to extrinsic, e.g. XYZ = ZY'X'' and vise versa. I made the assumption that the same applied to proper Euler angles, but now I'm doubting myself. I'm currently looking for information on this topic, and struggling to find a definitive answer. I'll comment again when I've got to the bottom of it.

**Comment by WestLangley** [comment]
So to clarify, all you need is for `Quaternion.setFromEuler( euler )` to support "proper" Euler angles. Correct?

And you do not need `Euler.setFromQuaternion( quaternion ) ` to support "proper" Euler angles. Correct?

And you do not need any of the other methods to support "proper" Euler angles?

**Comment by thezwap** [comment]
Correct. Here is my usage:
`var rotation = new THREE.Euler(rot1, rot2, rot3, 'ZXZ');
 var vectorXdash = new THREE.Vector3(1, 0, 0);
 vectorXdash.applyEuler(rotation);`

The applyEuler function in the Vector3, calls Quaternion.setFromEuler()

**Comment by WestLangley** [comment]
So, as I said previously, if were are to support the 6 additional Euler orders, we would have to support all possible orders in all relevant methods, and I think we would need a `ProperEuler` class to do so.

This is why something like `examples/js/math/Utils.js` could make sense, where we could define
```javascript
Quaternion.setFromProperEuler( a, b, c, order );
```
This method would implement your proposed code.

Or, we could create a static method:
```javascript
Quaternion.setQuaternionFromProperEuler( q, a, b, c, order );
```

**Comment by thezwap** [comment]
Oh, I see now. That sounds like it makes sense. That would work fine for me, and hopefully others.

**Comment by WestLangley** [comment]
/ping @XanderLuciano 
/ping @tentone  

Do you have any opinions on this proposed feature?

**Comment by XanderLuciano** [comment]
Overall, I'm in favor of adding support. 

I like the recommendation of using `THREE.MathUtils` for now as I don't expect it to be commonly needed. Having a `Quaternion.setFromProperEuler` and `Quaternion.setFromEuler` would likely cause confusion, require an extra class, and support for additional utilities for no real reason. The resultant `Quaternion` is all that's needed.

>I’ve started to see a few machine control apps built with threejs. See ncviewer.com as an example. That doesn’t do multi axis plane rotations yet as far as I know, but I’d be surprised if they weren’t on the roadmap.

NCViewer, to a certain degree, can do _some_ multi axis rotations:

image

But I'm getting around the limitation by using multiple `Quaternion.setFromAngleAxis` to manually set the rotation order so to speak. This means instead of just creating one Quaternion I'm creating up to 3 (and to be honest, that's mostly because it was easier than figuring out the math to do it properly).

This is also something that Autodesk worked on relatively recently with our post processor ([forum post about it](https://forums.autodesk.com/t5/hsm-post-processor-forum/calculating-euler-angles/td-p/6688480)), so to me it makes sense to add support for in Three.js as well. This way users won't have to figure out the math on their own if they want to work with Proper Euler angles.

Overall, it doesn't look like this would add much code and it would be useful to anyone working with simulating and driving multi axis CNC machines, multi axis 3D printers, and possibly even robotic arms (though I only have limited experience there).

**Comment by WestLangley** [comment]
Thanks, @XanderLuciano.

Let's do this: create `examples/js/utils/MathUtils.js`, and define
```javascript
THREE.MathUtils.setQuaternionFromProperEuler( q, a, b, c, order );
```
There is no need to include a testbed, but it would be nice to see a link to one with a DAT.GUI supporting `a`, `b`, `c`, and `order`.  A single `AxesHelper` for frame-of-reference, and one `BoxGeometry` controlled by the GUI would be sufficient to visually see that the code is working properly.

@thezwap Do you want to implement that? If not, I will be glad to.

**Comment by thezwap** [comment]
I'm keen to have a go and see how far I get. If I get completely stuck, I'll let you know and hand it over or ask for advice. Haven't used dat.gui before, but it looks fairly straightforward.

**Comment by WestLangley** [comment]
@thezwap I needed this for something else, so here is a stub to get you started: https://jsfiddle.net/7aLfd0zd/

screen shot 2018-03-12 at 11 55 03 pm

**Comment by thezwap** [comment]
Thanks. 

Gavin Bath
+64 27 5622583

**Comment by WestLangley** [comment]
@thezwap As I expected, for performance reasons, the code should be written like this, instead:
```javascript
q.set( c2 * s13, s2 * c1_3, s2 * s1_3, c2 * c13 );
```
rather than setting each component `.x`, .`y`, `.z`, `.w` separately. Otherwise, if `q` is the quaternion of an `Object3D` -- which it is in this case -- `Object3D.rotation` is updated four times.

/ping @mrdoob Ugh...

File: examples/js/utils/MathUtils.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        var cos = Math.cos;
+        var sin = Math.sin;
+
+        var c1 = cos(a / 2);
+        var c2 = cos(b / 2);
+        var c3 = cos(c / 2);
+
+        var s1 = sin(a / 2);
+        var s2 = sin(b / 2);
+        var s3 = sin(c / 2);
```
Comment (WestLangley) [inline examples/js/utils/MathUtils.js]: Four of these are not used.

**Comment by WestLangley** [comment]
>Is that a bad idea?

Yep. A quaternion that represents a rotation has unit length. The zero rotation is set by:  `q.set( 0, 0, 0, 1 );` 

You could leave the quaternion unchanged if the order is invalid. I do not think you need the intermediate variables `qx`, `qy`, etc.

**Comment by WestLangley** [comment]
@thezwap Thanks!

@mrdoob This can be merged. I'll handle the formatting issues.

**Comment by thezwap** [comment]
Thanks. That's my first pull request. Good to see how the process works.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**